Improve cache auto-configuration for Redis

ryonday opened the pull request:
Expose key prefix, TTL and null value settings for spring-data-redis' RedisCacheConfiguration in Spring .properties/yml configuration files.

Example:

```
spring.cache.redis.ttl=PT15M
spring.cache.redis.keyPrefix=foo
spring.cache.redis.useKeyPrefix=false
spring.cache.redis.cacheNullValues=false
```

Fixes gs-10795

ryonday commented:
Addresses [this](https://github.com/spring-projects/spring-boot/issues/10795)

snicoll commented on spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/cache/CacheProperties.java:
```diff
@@ -233,4 +242,72 @@ public void setConfig(Resource config) {
+	public static class Redis {
+
+		/**
+		 * The TTL for entries. By default, entries do not expire, a value of zero
```
Configuration key descriptions do not start with `The`, `A`, etc. Please look at [existing keys](https://docs.spring.io/spring-boot/docs/current/reference/html/common-application-properties.html) for inspiration

snicoll commented on spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/cache/CacheProperties.java:
```diff
@@ -233,4 +242,72 @@ public void setConfig(Resource config) {
+		 * The TTL for entries. By default, entries do not expire, a value of zero
+		 * indicates an eternal entry, and the given value will be converted to seconds..
+		 *
+		 * @see Duration#parse(CharSequence)
```
We don't make such `@see` link here. We use the javadoc to infer the description. Can you please remove them all?

snicoll commented on spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/cache/CacheProperties.java:
```diff
@@ -233,4 +242,72 @@ public void setConfig(Resource config) {
+		private boolean cacheNullValues = true;
+
+		/**
+		 * The non-null Redis key prefix to use instead of the default one.
```
I am not sure what you mean by "non-null" here.

snicoll commented on spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/cache/CacheProperties.java:
```diff
@@ -233,4 +242,72 @@ public void setConfig(Resource config) {
+			return this;
+		}
+
+		public Optional<String> getKeyPrefix() {
```
Configuration properties must be JavaBean properties (the type must match) and we don't support `Optional` here.

snicoll commented on spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/cache/RedisCacheConfiguration.java:
```diff
@@ -57,13 +59,38 @@
 
 	@Bean
 	public RedisCacheManager cacheManager(RedisConnectionFactory redisConnectionFactory) {
+
```
Please refrain from making unnecessary formatting changes.

snicoll commented on spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/cache/RedisCacheConfiguration.java:
```diff
@@ -57,13 +59,38 @@
 		RedisCacheManagerBuilder builder = RedisCacheManager
-				.builder(redisConnectionFactory);
+				.builder(redisConnectionFactory).cacheDefaults(getConfiguration());
+
```
Idem

snicoll commented on spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/cache/RedisCacheConfiguration.java:
```diff
@@ -57,13 +59,38 @@
 		if (!cacheNames.isEmpty()) {
 			builder.initialCacheNames(new LinkedHashSet<>(cacheNames));
 		}
+
```
Idem

ryonday commented:
Thanks so much for the detailed feedback, @snicoll; I hope that the follow-up commit has addressed your concerns! 

I'm happy to be able to give back in some small way to a project that I've used so much throughout my career.

snicoll commented:
@ryonday Thanks for the feedback. They did partially and I had an extra polish pass. It's now merge in `master`.

ryonday commented:
Thanks so much for the polish pass, I have a few more small PR ideas and it helps to have a template for the preferred conventions and style to make it easier on reviewers! 

Cheers!
